optimizely/odp/odp_event_manager.py

```diff
@@ -0,0 +1,186 @@
+from optimizely.helpers.enums import OdpEventManagerConfig, Errors
+
+
+if version_info < (3, 8):
+    from typing_extensions import Final
+else:
+    from typing import Final  # type: ignore
+
+
+class Signal:
+    '''Used to create unique objects for sending signals to event queue.'''
+    pass
+
+
+class OdpEventManager:
+    """
+    Class that sends batches of ODP events.
+
+    The OdpEventManager maintains a single consumer thread that pulls events off of
+    the queue and buffers them before the events are sent to ODP.
+    """
+
+    _SHUTDOWN_SIGNAL: Final = Signal()
+    _FLUSH_SIGNAL: Final = Signal()
+
+    def __init__(
+        self,
+        odp_config: OdpConfig,
+        logger: Optional[_logging.Logger] = None,
+        api_manager: Optional[ZaiusRestApiManager] = None
+
+    ):
+        """ OdpEventManager init method to configure event batching.
+
+        Args:
+            odp_config: ODP integration config.
+            logger: Optional component which provides a log method to log messages. By default nothing would be logged.
+            api_manager: Optional component which sends events to ODP.
+        """
+        self.logger = logger or _logging.NoOpLogger()
+        self.zaius_manager = api_manager or ZaiusRestApiManager(self.logger)
+        self.odp_config = odp_config
+        self.event_queue: Queue[OdpEvent | Signal] = Queue(OdpEventManagerConfig.DEFAULT_QUEUE_CAPACITY)
+        self.batch_size = OdpEventManagerConfig.DEFAULT_BATCH_SIZE
+        self._current_batch: list[OdpEvent] = []
+        self.executor = Thread(target=self._run, daemon=True)
+
+    @property
+    def is_running(self) -> bool:
+        """ Property to check if consumer thread is alive or not. """
+        return self.executor.is_alive()
+
+    def start(self) -> None:
+        """ Starts the batch processing thread to batch events. """
+        if self.is_running:
+            self.logger.warning('ODP event processor already started.')
+            return
+
+        self.executor.start()
+
+    def _run(self) -> None:
+        """ Triggered as part of the thread which batches odp events or flushes event_queue and blocks on get
+        for flush interval if queue is empty.
+        """
+        try:
+            while True:
+                item = self.event_queue.get()
+
+                if item == self._SHUTDOWN_SIGNAL:
+                    self.logger.debug('Received ODP event shutdown signal.')
+                    self.event_queue.task_done()
+                    break
+
+                if item is self._FLUSH_SIGNAL:
+                    self.logger.debug('Received ODP event flush signal.')
+                    self._flush_batch()
+                    self.event_queue.task_done()
+                    continue
+
+                if isinstance(item, OdpEvent):
+                    self._add_to_batch(item)
+                    self.event_queue.task_done()
+
+        except Exception as exception:
```

Comment:
I know existing BatchEventProcessor uses this catch-all. I wonder we could use a less generic class of exceptions, however I can see it wouldn't be straightforward to isolate them (issues with signaling, flushing, adding to batch,...). Maybe capturing all is desirable here nevertheless.

Reply:
Since exceptions in child threads don't propagate to the main thread, we should keep this exception. I also added a flag so that the main thread can throw errors on subsequent attempted send events.